`storage`: do not recover `segment`s with unordered offsets

WillemKauf opened the pull request:
This PR should address the failure seen in https://redpandadata.atlassian.net/browse/CORE-12745.

Up till now, when we encounter the case in which `segment`s being recovered have `prev.offsets().get_dirty_offset() >= s.offsets().get_base_offset()`, we assume the `segment_index` is corrupted (i.e the offsets are incorrect), and use a `log_replayer` to ensure correctness of the `segment` offsets.

However, after this recovery process, we do nothing to ensure that the resulting `segment_set` is composed of monotonically increasing offsets, or rather, that the assumed "incorrect" offsets weren't actually correct to begin with.

This scenario can easily occur with adjacent merge compaction, in which the replacement `segment` has been produced from the concatenation operation, but the `redpanda` broker crashes/goes down before removing all of the `segment`s that were part of the concatenation. This race condition was possible before, and has only been made more likely with the [recent enhancements made to adjacent segment compaction](https://github.com/redpanda-data/redpanda/pull/25953) to support the concatenation of more than two `segment`s.

To prevent these `segment`s from being recovered, perform the same offset check AFTER recovery of potentially corrupted `segment`s has occurred. If the same condition `prev.offsets().get_dirty_offset() >= s.offsets().get_base_offset()` is `true`, it _very likely_ indicates that there was no corruption at all, and that there are redundant `segment`s left on disk from an incomplete adjacent segment compaction (incomplete meaning the `redpanda` broker went down before all now redundant `segment`s were removed- it is guaranteed at this point that the replacement `segment` is corruption-free and holds all of the necessary data from the redundant `segment`s).

To prevent any false positive cases or incorrect data loss scenarios (similar to the case in which we cannot recover a `segment`), rename these redundant `segment` files with a `.ignore_have_newer` suffix for use in the case of necessary manual recovery.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

WillemKauf commented:
Force push to:
* `clang-format`

WillemKauf commented on src/v/storage/segment_set.cc:
```diff
@@ -346,7 +349,37 @@ static ss::future<segment_set> unsafe_do_recover(
+                    vlog(
+                      stlog.info,
+                      "Base offset {} of segment {} is < dirty offset {} of "
+                      "previous segment {} after recovery. This is very likely "
```
I am ready to make the claim that _any_ segment that falls under this category (i.e a redundant segment that was not removed by adjacent merge compaction) could instead be safely removed instead of renamed (`ss::remove_file()` instead of `ss::rename_file()`) here.

I'm also ready to make the claim that there are no cases in which non-monotonic offsets here are a valid state for a `segment_set`, making ALL segments that fit this criteria eligible for removal instead of renaming.

However, in my conservative/defensive coding soul, I still wrote this as a file rename instead of a file removal. 

I am wondering if anyone else with a less nervous coding style is in favour of removal over rename here.

Lazin replied:
maybe it makes sense to add a check that will guarantee that after the rename/removal we will still have monotonic offset range? isn't it the case that prev.dirty will be equal to base_offset of the next in this case?

WillemKauf replied:
Done in new function `maybe_create_contiguous_segment_set()`

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--66632148-5282-4770-8967-bef6cae270bf-->
<!--TEST RESULTS-->
<!--f01a38ca-cc99-4b8d-a9ce-603893fbb8e8-->
<!--TEST RESULTS-->
<!--31fb04f8-e256-4e6c-a668-9d5ad4381225-->
<!--TEST RESULTS-->
<!--f5476b43-1b9d-4e1f-a2de-30bcb92c8b68-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68678">build#68678</a></summary>


| test_class               | test_method             | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest             | test_connection_flicker | {"cloud_storage_type": 1}                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68678#0197f12d-3567-47e1-aaf7-d23e7eafc7cf | FLAKY         | 17/21    | upstream reliability is '99.71751412429379'. current run reliability is '80.95238095238095'. drift is 18.76513 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations    | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68678#0197f128-5f61-4b32-a7d7-cc7299f2999a | FLAKY         | 20/21    | upstream reliability is '96.7741935483871'. current run reliability is '95.23809523809523'. drift is 1.5361 and the allowed drift is set to 50. The test should PASS    |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68755">build#68755</a></summary>


| test_class                  | test_method                | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:----------------------------|:---------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68755#0197f584-ebcb-48ab-bfd6-ad9bd94b76c7 | FLAKY         | 20/21    | upstream reliability is '97.81659388646288'. current run reliability is '95.23809523809523'. drift is 2.5785 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68827">build#68827</a></summary>


| test_class                         | test_method                   | test_arguments                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------------|:------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest       | test_cloud_storage            | {"cleanup_policy": "compact,delete"}                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68827#0197f9fc-a670-4aa0-badc-565866a0f521 | FLAKY         | 19/21    | upstream reliability is '94.9468085106383'. current run reliability is '90.47619047619048'. drift is 4.47062 and the allowed drift is set to 50. The test should PASS  |
| ControllerLogLimitMirrorMakerTests | test_mirror_maker_with_limits |                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68827#0197f9fc-a672-4132-bc20-88e44fe48a3f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| DatalakeE2ETests                   | test_json_schema_unicode      | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"}                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68827#0197fa00-8c49-4669-9d8e-a76e14ad776f | FLAKY         | 20/21    | upstream reliability is '98.01980198019803'. current run reliability is '95.23809523809523'. drift is 2.78171 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest           | test_node_operations          | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68827#0197f9fc-a674-45be-9d70-7f4c20604f52 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68920">build#68920</a></summary>


| test_class            | test_method                               | test_arguments                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------|:------------------------------------------|:---------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePoolMigrationTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68920#019809bc-d668-4f62-b185-48b2e11d61b5 | FLAKY         | 20/21    | upstream reliability is '99.46236559139786'. current run reliability is '95.23809523809523'. drift is 4.22427 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

nvartolomei commented:
@WillemKauf the way you describe the problem makes me think that a much simpler and direct fix here would be enough. Do we have both start and end offsets of the segments at that point? Then, can't we run an algorithm to identify the exact overlaps in the segments and remove the ones that are being "overlapped" by a larger segment?

I.e. if I have segment 0-100 and segments 0-50, 50-100 then and only then remove the 0-50, 50-100.

For other cases error out? The current code, on quick glance, can mask some other bugs too including corruption and data loss.

nit: `.tombstone` -> `.ignore-have-newer` or something like that. tombstone is too generic. ignore-have-newer encodes the cause

nvartolomei commented:
Not directly related to this PR but shouldn't we also have a check in there that there are no gaps in segment set? Our lower_bound method on segment set iirc does not tolerate gaps.

WillemKauf commented:
>Then, can't we run an algorithm to identify the exact overlaps in the segments and remove the ones that are being "overlapped" by a larger segment?
> I.e. if I have segment 0-100 and segments 0-50, 50-100 then and only then remove the 0-50, 50-100.

For your example, segments `[0-100],[0-50],[50-100]`, the code currently works by _assuming_ removal of `[0-50]`, `[50-100]` will maintain continuity in the offset space. It doesn't identify anything "exactly", but it does rely on the fact that the post recovery segment collection/offsets are correct. Are these assumptions what you have qualms with?

> For other cases error out? The current code, on quick glance, can mask some other bugs too including corruption and data loss.

I.e the case in which gaps do exist in the segment set? What does error'ing out look like in this case? `redpanda` refusing to start, silent removal of data via a rename to `.not_contiguous_segment`, etc?

> nit: .tombstone -> .ignore-have-newer or something like that. tombstone is too generic. ignore-have-newer encodes the cause

Ack, I like this much better too :)

nvartolomei commented:
@WillemKauf 
> Are these assumptions what you have qualms with?

Yes. You are making an assumption that if base <= prev.dirty then a specific problem occurred (the one you described). However, in the long run, that might not be necessarily the case. You are also assuming that removing these segments will lead to a valid log. That is true only for the case you described. But again, the code nowhere check that this is actually what have happened.

> I.e the case in which gaps do exist in the segment set? What does error'ing out look like in this case? redpanda refusing to start, silent removal of data via a rename to .not_contiguous_segment, etc?

Good question. I don't know what the solution looks like there... Inserting a faux empty segment? Refusing to start the partition might be an option too.

WillemKauf commented:
>You are also assuming that removing these segments will lead to a valid log. That is true only for the case you described. But again, the code nowhere check that this is actually what have happened.

Added a check here which will refuse to start the partition via thrown exception if a non-contiguous offset span post recovery is encountered in the `segment_set`.

Do you think this is enough to make the original changes viable?

WillemKauf commented:
Oops. Looks like this condition already got triggered in a fixture test:

```
unknown location(0): fatal error: in "test_upload_with_gap_blocked": non_contiguous_log_exception: Offsets found in recovered segments are non-contiguous after segment recovery. This could indicate corruption of segment data.
```

WillemKauf commented on src/v/cluster/archival/tests/ntp_archiver_test.cc:
```diff
@@ -1447,110 +1447,6 @@ FIXTURE_TEST(test_manifest_spillover, archiver_fixture) {
 }
 
-// NOLINTNEXTLINE
-FIXTURE_TEST(test_upload_with_gap_blocked, archiver_fixture) {
```
Removed test as it is ill-posed with non-contiguous local `segment`s.

andrwng replied:
Is the test not useful? Wondering if there are other conditions that could set the log up in a way that preserves the spirit of this test? Or if we should add a test-only mode that allows gaps or something.

nvartolomei replied:
The test needs to be rewritten to exercise the code path we are interested in while also not violating local storage invariants.

Lazin replied:
The test is not useful IMO.
It creates a gap in the manifest and checks that compacted reupload can fill the gap. I think we shouldn't get into this situation to begin with. The test was added long time ago. I have no idea what was the motivation behind it.

andrwng commented on src/v/storage/segment_set.cc:
```diff
@@ -349,7 +349,37 @@ static ss::future<segment_set> unsafe_do_recover(
             good.emplace_back(std::move(s));
         }
-        return segment_set(std::move(good));
+
+        auto seg_set = segment_set(std::move(good));
```
Could we add a unit test for recover_segments() that exercises this code?

WillemKauf replied:
Added `segment_set_test.cc` which unit tests new added function `maybe_create_contiguous_segment_set()`, and existing test `TestAdjacentCompactionSimulateCrashDuringFileRemoval` in `compaction_e2e_test.cc` also stresses this code path.

andrwng commented on src/v/storage/segment_set.cc:
```diff
@@ -349,7 +349,37 @@ static ss::future<segment_set> unsafe_do_recover(
+                      s.reader().filename(),
+                      s.reader().filename() + ".ignore_have_newer")
+                      .get();
+                    it = seg_set.erase(it);
```
Is there a particular rationale for why we're erasing `it` and not `prev`? Is there an ordering property in segment_set that we're relying on to ensure the one we're deleting is actually the newer one? E.g. should we always keep the one with the larger offset range?

WillemKauf replied:
`prev` is now guaranteed to be the `segment` with a greater offset range and valid offset continuity, as we now sort by both ascending order w/r/t base offset and descending order w/r/t dirty offset in `segment_ordering`.

WillemKauf committed:
`storage`: add `maybe_create_contiguous_segment_set()`

Up till now, when we encounter the case in which `segment`s being recovered
have `prev.offsets().get_dirty_offset() >= s.offsets().get_base_offset()`,
we assume the `segment_index` is corrupted (i.e the offsets are incorrect),
and use a `log_replayer` to ensure correctness of the `segment` offsets.

However, after this recovery process, we do nothing to ensure that the
resulting `segment_set` is composed of monotonically increasing offsets,
or rather, that the assumed "incorrect" offsets weren't actually correct to
begin with.

This scenario can easily occur with adjacent merge compaction, in which the
replacement `segment` has been produced from the concatenation operation,
but the `redpanda` broker crashes/goes down before removing all of the
`segment`s that were part of the concatenation. This race condition was
possible before, and has only been made more likely with the recent
enhancements made to adjacent segment compaction to support the concatenation
of more than two `segment`s.

In order to generate a contiguous, non-overlapping `segment_set` from the
`segment`s recovered and those left on disk, add
`maybe_create_contiguous_segment_set()`. This function takes a greedy search
approach in an attempt to prevent any redundant/duplicate data from being
managed in the `log` after `segment_set` recovery is performed. If the
greedy approach fails to create a `segment_set` under these constraints, the
function returns `std::nullopt` and the original deque of `segment`s
will be used instead to build the `log` (i.e the old behavior is preserved).

WillemKauf committed:
`storage`: use `maybe_create_contiguous_segment_set()` during `segment` recovery

As mentioned, if this function fails to create a contiguous non-overlapping
set of `segment`s to represent the `log` with, the original `deque` of `segment`s
is returned instead (i.e the old behavior is preserved).

WillemKauf commented:
Force push to:
* add descending dirty offset ordering to `segment_ordering`
* Rework PR with a new function, `maybe_create_contiguous_segment_set()`. This is used during `segment` recovery to try to create a contiguous, non-overlapping `segment_set` out of those available using a greedy approach (`segment`s which span the highest offset range are selected), and checks that the end result is valid. If it is not, the function returns `std::nullopt` and the original `deque` of `segment`s is used for the `log` (i.e the old behavior is preserved).
* add `segment_set_test` for test coverage of `maybe_create_contiguous_segment_set()`

WillemKauf commented on src/v/storage/segment_set.cc:
```diff
@@ -423,6 +423,11 @@ static ss::future<segment_set> unsafe_do_recover(
             good.emplace_back(std::move(s));
         }
+
+        auto seg_set_opt = maybe_create_contiguous_segment_set(good).get();
```
`good` is copied (pass by value) here.

andrwng replied:
nit: please add this as a comment and explain why it's important so readers who don't see this PR aren't tempted to change it

WillemKauf replied:
Added comment.

WillemKauf commented:
Force push to:
* Add check for empty `deque` in `maybe_create_contiguous_segment_set()` to avoid bad dereference

WillemKauf commented:
One more force push to:
* reduce number of passes over `segment_set` in `maybe_create_contiguous_segment_set` from 2 to 1.

andrwng commented on src/v/storage/segment_set.cc:
```diff
@@ -27,13 +27,19 @@
+        // Order by ascending base offset and descending dirty offset.
+        return std::forward_as_tuple(
+                 o1.get_base_offset(), o2.get_dirty_offset())
+               < std::forward_as_tuple(
```
>which is the desired behavior (`segment`s which span the highest offset range
>are ordered first).

I understand we want it for this change, but is this desired behavior in general? Reordering all segment_sets seems like a big change for what I thought could be a fairly targeted fix

If this is actually needed for correctness in the read path and elsewhere, could you leave a comment explaining why the ordering the important? Even with context of this bug, it's not intuitive to me why this aspect of segment ordering is important for all callers.

WillemKauf replied:
I agree. Removed this alteration to `segment_ordering` and made a direct call within `maybe_create_contiguous_segment_set()` to `std::sort()` with this new definition.

Unfortunately this means we will sort the `segment`s twice in the happy path now.

andrwng commented on src/v/storage/segment_set.cc:
```diff
@@ -197,6 +199,83 @@ is_last_segment(segment* s, std::optional<ss::sstring> last_clean_segment) {
+                        ->offsets()
+                        .get_dirty_offset();
+    segment_set::underlying_t ignored_segs;
+    for (auto it = seg_set.begin(); it != seg_set.end();) {
+        if (it != seg_set.begin()) {
+            auto& s = *it;
+            auto& prev = *std::prev(it);
```
nit: can we start from `std::next(seg_set.begin())` and avoid this extra scope?

WillemKauf replied:
Yep, changed this, but also added an early return for

```cpp
    if (segs.size() < 2) {
        co_return std::nullopt;
    }
```

otherwise `std::next()` would be an invalid dereference.

andrwng commented on src/v/storage/segment_set.cc:
```diff
@@ -197,6 +199,83 @@ is_last_segment(segment* s, std::optional<ss::sstring> last_clean_segment) {
+        if (it != seg_set.begin()) {
+            auto& s = *it;
+            auto& prev = *std::prev(it);
+            if (
```
Given we're relying on ordering of the underlying segment_set, could we create a std::deque of segments in this method and sort them exactly how we need, rather than changing the sort order of segment_set and relying on it implicitly

If you choose to keep it as is, could you add a comment explaining that we are relying on a specific ordering of the last offsets of these segments?

WillemKauf replied:
Changed as mentioned above.

andrwng commented on src/v/storage/segment_set.cc:
```diff
@@ -197,6 +199,83 @@ is_last_segment(segment* s, std::optional<ss::sstring> last_clean_segment) {
+                continue;
+            }
+            // At this point we have greedily selected the segment `s`- ensure
+            // offset continuity in the same pass
```
nit: Looking at this code fresh, this comment about "same pass" seems out of place. Is "same pass" alluding to another implementation that had multiple passes?

WillemKauf replied:
Removed "same pass" comment.

andrwng commented on src/v/storage/segment_set.cc:
```diff
@@ -197,6 +199,83 @@ is_last_segment(segment* s, std::optional<ss::sstring> last_clean_segment) {
+        ++it;
+    }
+
+    if (seg_set.empty()) {
+        co_return std::nullopt;
+    }
```
Is this dead code, given we don't ever erase the first segment?

WillemKauf replied:
Yes, this path shouldn't ever be taken, but I felt more confident with a sanity check ahead of raw dereferences to `.front()` and `.back()`. Removed it anyways.

andrwng commented on src/v/storage/segment_set.cc:
```diff
@@ -197,6 +199,83 @@ is_last_segment(segment* s, std::optional<ss::sstring> last_clean_segment) {
+
+    // Offset span needs to be preserved.
+    if (!(seg_set.front()->offsets().get_base_offset() == min_offset
+          && seg_set.back()->offsets().get_dirty_offset() == max_offset)) {
```
nit: think its worth a warning? or debug log or something?

WillemKauf replied:
Adding `warn` level log line here.

andrwng commented on src/v/storage/segment_set.cc:
```diff
@@ -197,6 +199,83 @@ is_last_segment(segment* s, std::optional<ss::sstring> last_clean_segment) {
+    }
+
+    for (auto& s : ignored_segs) {
+        co_await ss::rename_file(
+          s->reader().filename(),
+          s->reader().filename() + ".ignore_have_newer");
```
If you're intent on not deleting the files, we should remove them in `log_manager::remove()` so they don't block partition removal

WillemKauf replied:
Do `.cannotrecover` files currently block removal?

andrwng replied:
I'm not sure, but my guess is yes. My guess is also that they're significantly more rare and caused by bugs, whereas `ignore_have_newer` files are expected in normal operation of a cluster.

WillemKauf replied:
Added considerations for `.cannotrecover` and `.ignore_have_newer` files in `log_manager::remove()`.

WillemKauf replied:
> are expected in normal operation of a cluster.

Maybe the log line for removal of these `.ignore_have_newer` files should be lower than `warn` then.

andrwng reviewed (commented):
Nice updates!

WillemKauf committed:
`storage`: clean up `.cannotrecover` and `.ignore_have_newer` files in `log_manager::remove()`

Otherwise, these files could potentially block `partition` removal when
attempting to remove `ntp_dir`.

nvartolomei commented on src/v/storage/log_manager.cc:
```diff
@@ -854,15 +854,25 @@ ss::future<> log_manager::remove(model::ntp ntp) {
               // can be problematic for users, as it needlessly consumes space.
               // Log verbosely to make it easier to catch.
+              leftover_file_opt = "staging";
+          } else if (de.name.ends_with(".cannotrecover")) {
```
This is outdated code? Don't see a codepath creating `.cannotrecover` files. In case there is/should be one I suggest naming files with `.ignore_` prefix so that then we can have code that cleanups everything based on prefix rather than maintaining special cases.

WillemKauf replied:
`.cannotrecover` files are produced here: https://github.com/redpanda-data/redpanda/blob/9921f72cbbab884aa4300ba98a33dd25309e030a/src/v/storage/segment_set.cc#L328-L337, when offset corruption is detected and the `log_replayer` is unable to recover a file.

nvartolomei commented on src/v/storage/log_manager.cc:
```diff
@@ -854,15 +854,25 @@ ss::future<> log_manager::remove(model::ntp ntp) {
               // proceed with removal, but it points to a missing cleanup which
               // can be problematic for users, as it needlessly consumes space.
               // Log verbosely to make it easier to catch.
+              leftover_file_opt = "staging";
```
Why do we need `leftover_file_opt`? The filepath already has all this information so what it helps with?

WillemKauf replied:
There are 3 code paths here (`staging`, `ignore_have_newer` and `cannotrecover`) which should all have their file logged & removed, so I felt that writing it this way with an `std::optional` was clean, logic/branch wise.

WillemKauf replied:
Altered with code recommendations from `slack`

nvartolomei commented on src/v/storage/tests/segment_set_test.cc:
```diff
@@ -0,0 +1,250 @@
+// Copyright 2025 Redpanda Data, Inc.
```
nit: squash test with the code which changes the behavior.

WillemKauf replied:
Squashed

nvartolomei commented on src/v/storage/segment_set.cc:
```diff
@@ -437,6 +437,14 @@ static ss::future<segment_set> unsafe_do_recover(
+        if (seg_set_opt.has_value()) {
+            return std::move(seg_set_opt).value();
+        }
         return segment_set(std::move(good));
```
> if this function fails to create a contiguous non-overlapping set of `segment`s to represent the `log` with, the original `deque` of `segment`s is returned instead (i.e the old behavior is preserved).

So we are accepting running with a segment set which does not make sense and results pretty much in undefined behavior? Why this instead of failing to start the partition and asking the operator to intervene?

WillemKauf replied:
@andrwng and I discussed this PR as a whole and decided that without a better solution (i.e a non-greedy search algorithm that creates the contiguous segment set) it would be best to ONLY target the case we are worried about (i.e adjacent segment compaction) and fall back to the old behavior if the fix for that specific case didn't work.

andrwng replied:
Yeah, my rationale was that the edge case Willem has been concerned about is actually pretty narrow -- unclean shutdown during compaction. That seemed easy enough to provide a targeted fix with somewhat of a guarantee that we don't regress anything else.

I'm not against doing the broader change too, but I don't think it needs to block this change, and may warrant a bit more testing and bake time
